### ISSUE TITLE
How can i implement a GraphSAGE ?

### ISSUE THREAD

**Issue by 7ianwo0o** [rusty1s/pytorch_geometric#1050]
## ❓ Questions & Help

I just want implement a GraphSAGE.
I try to do it but the results are confusing. So i want to know if my implementation is right.
Here is my code:
At first i want to load my spilts and creat a subgraph for inductively training.
```
train_idx,val_idx,test_idx = load_spilts(dsName,times)
edge_index_inductive, _ = subgraph(train_idx,data.edge_index)
```

and create a new Data
`data_new = Data(x=data.x,edge_index=edge_index_inductive,y=data.y,
                train_mask=train_mask,val_mask=val_mask,
                test_edge_index=data.edge_index,
                test_mask=test_mask,
                test_num_nodes=1000)`
My model is as follows:
```
class graphSAGE_paper_K_2(torch.nn.Module):
    def __init__(self,dataset,features_dimension):
        super(graphSAGE_paper_K_2,self).__init__()
        self.conv1 = SAGEConv(features_dimension, 16)
        self.conv2 = SAGEConv(16,dataset.num_classes)

    def forward(self,x,edge_index):
        x, edge_index = x,edge_index
        x = self.conv1(x, edge_index)
        x = F.relu(x)
        x = F.normalize(x, p=2, dim=-1)
        x = self.conv2(x, edge_index)
        x = F.relu(x)
        x = F.normalize(x, p=2, dim=-1)
        return F.log_softmax(x, dim=1)
```
and potimizer parameter:
`device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model, data = model.to(device), data_new.to(device)
optimizer = torch.optim.Adam(model.parameters(),lr=0.001, weight_decay=0)`

train and test:
#------------------------------train function-----------------------------
```
def train():
    model.train()
    optimizer.zero_grad()
    F.nll_loss(model(data_new.x,data_new.edge_index)[data_new.train_mask],data_new.y[data_new.train_mask]).backward()
    optimizer.step()
```
#----------------------------------------------------------
----------- TEST FUNCTION -----------------------------------------
```
def test():
    model.eval()
    pred_train = model(data_new.x,data_new.edge_index)[data_new.train_mask].max(1)[1]
    pred_test = model(data_new.x,data_new.test_edge_index)[data_new.test_mask].max(1)[1]
    pred_val = model(data_new.x,data_new.test_edge_index)[data_new.val_mask].max(1)[1]
    pred_labels = pred_test.tolist()
    true_labels = data_new.y[data_new.test_mask].tolist()
    train_acc = get_train_acc(data_new,pred_train)
    test_acc = get_test_acc(data_new,pred_test)
    val_acc = get_val_acc(data_new,pred_val)
    f1_marco = get_marco_f1(data_new,pred_labels,true_labels,lables_list)
    f1_micro = get_micro_f1(data_new,pred_labels,true_labels,lables_list)
    return train_acc,val_acc,test_acc,f1_marco,f1_micro

```
#--------------------------TRAIN PROCESS-----------------------------------------------
```
best_val_acc = best_val_acc = 0
for epoch in range(1,201):
    train()
    train_acc, val_acc, tmp_test_acc,tmp_marco,tmp_micro = test()
    if val_acc > best_val_acc:
        best_val_acc = val_acc
        test_acc = tmp_test_acc
        marco = tmp_marco
        micro = tmp_micro
    log = 'Epoch: {:03d}, Train: {:.4f}, Val: {:.4f}, Test: {:.4f},marco: {:.4f},micro: {:.4f}'
    print(log.format(epoch, train_acc, best_val_acc, test_acc,marco,micro))`

```
but the resutls on dataset Cora are so bad:
```
Epoch: 196, Train: 0.5500, Val: 0.3240, Test: 0.3410,marco: 0.2689,micro: 0.3410

Epoch: 197, Train: 0.5500, Val: 0.3240, Test: 0.3410,marco: 0.2689,micro: 0.3410
Epoch: 198, Train: 0.5500, Val: 0.3240, Test: 0.3410,marco: 0.2689,micro: 0.3410
Epoch: 199, Train: 0.5500, Val: 0.3240, Test: 0.3410,marco: 0.2689,micro: 0.3410
Epoch: 200, Train: 0.5500, Val: 0.3240, Test: 0.3410,marco: 0.2689,micro: 0.3410
```
so if i did something wrong?

**Comment by rusty1s** [comment]
Try to remove
```python
x = F.relu(x)
x = F.normalize(x, p=2, dim=-1)
```
before the `log_softmax` call.

**Comment by 7ianwo0o** [comment]
> Try to remove
> 
> ```python
> x = F.relu(x)
> x = F.normalize(x, p=2, dim=-1)
> ```
> 
> before the `log_softmax` call.

Yeap . you are right. it gets better . But the marco score is still about 0.55.  I am not sure if it is acceptable.

![Jietu20200321-下午100207@2x](https://user-images.githubusercontent.com/24668785/77236532-b8b4be80-6bbf-11ea-89e3-c0baadc49ef9.jpg)
But I am confused why should I remove these two steps. Coz I find in the paper, after each GraphSAGE layer i need to do activation and normalization. Am I right?

**Comment by rusty1s** [comment]
Well, you can use Relu and normalization in intermediate layers, but it does not make sense to restrict your features before inputting them to a softmax. You want your model to be able to push wrong targets to minus infinity, and correct target to infinity.

**Comment by 7ianwo0o** [comment]
Well Thanks for your help!
